[Merged by Bors] - Smesher API + GPU/Tree-free PoST integration

moshababo opened the pull request:
Implemented the new [Smesher API](https://github.com/spacemeshos/api/blob/master/proto/spacemesh/v1/smesher.proto) (instead of the previous Mining API), while integrating GPU/Tree-free PoST.
This PR is based on https://github.com/spacemeshos/post/pull/41, which need to be reviewed first, hence this description won't repeat what was already been specified there.

Main remarks:

* Created `PostSetupManager` for implementing the new Post setup API. The previous `PostClient` was removed, and setup functionalities previously implemented by the ATX builder were removed as well. 
* Refactored the ATX builder to comply with the new Smeshing API. Its dependency with `PostSetupManager` is only regarding the setup completion status and generation of proofs. 
* In addition to the raw Post, the NIPost now contains the Post metadata as well, that the proof is expected to be verified upon during the syntactic validation. Some of the metadata values are expected to be derived from consensus, hence should be validated in the contextual validation, but it is currently still being derived from the node configuration, so it’s still taking place in the syntactic validation (which will need to remain anyway for validating min/max boundaries).
* Added `PostSetupOpts` config. It is used for initiating a new data creation session - either via the Smesher API, or on node start. 
* Renamed the ATX’s `commitment` as `initialPost`. As discussed with @tal-m, in the tree-free PoST the proof is *not* a commitment to the data.
* Changed NIPoST’S PoET challenge to contain the initial Post indices, instead of its Merkle root, which is no longer applicable. The reason is the same - it is expected to be known at the time the challenge was constructed. 
* Renamed “nipst” as “NIPost” throughout the codebase (closing https://github.com/spacemeshos/go-spacemesh/issues/1371).
* Some of the old API code and test-code was commented out or disabled due to compile-time errors caused to the new defined interfaces. It is expected to be fully removed soon.

moshababo committed:
Smesher API + GPU/Tree-free PoST integration

moshababo committed:
Merge fix

moshababo committed:
Merge branch 'develop' into smesher_api

# Conflicts:
#	cmd/node/node.go

lrettig commented on api/grpcserver/smesher_service.go:
```diff
@@ -25,40 +27,34 @@ func (s SmesherService) RegisterService(server *Server) {
 func (s SmesherService) StartSmeshing(ctx context.Context, in *pb.StartSmeshingRequest) (*pb.StartSmeshingResponse, error) {
 	log.Info("GRPC SmesherService.StartSmeshing")
 
+	// TODO(moshababo): remove DataDir and CommitmentSize from the request proto definition.
```
Please open an issue for this here: https://github.com/spacemeshos/api/issues

moshababo replied:
https://github.com/spacemeshos/api/issues/114

lrettig commented on api/grpcserver/smesher_service.go:
```diff
@@ -25,40 +27,34 @@ func (s SmesherService) RegisterService(server *Server) {
 	log.Info("GRPC SmesherService.StartSmeshing")
 
+	// TODO(moshababo): remove DataDir and CommitmentSize from the request proto definition.
+	// TODO(moshababo): check why JSON request via HTTP gateway doesn't decode `coinbase` properly
```
This is a bit tricky since it's expecting bytes, not a string. Possibly useful: https://github.com/spacemeshos/api/issues/109

moshababo replied:
I've used a similar encoding (and inspected the encoding used in responses), but it doesn't work well for `coinbase` field for some reason (unlike `account_id`, both of the same type). 
We also have examples of this in https://github.com/spacemeshos/cliwallet.

My format:
```
{
    "coinbase": {
        "address": "QWxhZGRpbjpvcGVuIHNlc2FtZQ=="
    }
}
```
I've tried it a while ago. Will give it another shot before merging, just for clients' well-being.

lrettig replied:
That's really strange. `coinbase` in the SmesherService is of type `account_id` which is just an address passed as an array of bytes, as elsewhere in the API. Do you have sample curl commands, one that works somewhere else and one that fails here?

avive replied:
It should work ok. CLIWallet uses binary data for all addresses in GRPC calls and it works as expected. There's no need to base64 binary data in grpc - only when working higher-level via the json-proxy as there's no binary data in json.

lrettig commented on api/grpcserver/smesher_service.go:
```diff
@@ -78,21 +76,16 @@ func (s SmesherService) StopSmeshing(ctx context.Context, in *pb.StopSmeshingReq
-	}
-	return &pb.CoinbaseResponse{AccountId: &pb.AccountId{Address: addr.Bytes()}}, nil
+	addr := s.smeshing.Coinbase()
+	return &pb.CoinbaseResponse{AccountId: &pb.AccountId{Address: addr[:]}}, nil
```
For my own edification, what does `addr[:]` do? Is it the same as `addr.Bytes()`?

moshababo replied:
Yes, it's the same. I guess I changed it after I rewrote this section, and didn't notice the `Bytes()` alias, which is probably preferable. Fixed.

lrettig commented on api/grpcserver/smesher_service.go:
```diff
@@ -125,31 +119,112 @@ func (s SmesherService) SetMinGas(ctx context.Context, in *pb.SetMinGasRequest)
+	for i, p := range providers {
+		hs, err := p.Benchmark()
+		if err != nil {
+			return nil, err
```
Please don't return `err` directly here. The pattern we've been using is to pass the raw error to `log.Error` and then return `status.Error(codes.Internal, ...)` with a short description of the error fit for external consumption.

lrettig commented on api/grpcserver/smesher_service.go:
```diff
@@ -125,31 +119,112 @@ func (s SmesherService) SetMinGas(ctx context.Context, in *pb.SetMinGasRequest)
-	return nil, status.Errorf(codes.Unimplemented, "this endpoint is not implemented")
+
+	if in.Data == nil {
+		return nil, status.Errorf(codes.InvalidArgument, "`Data` must be provided")
```
```suggestion
		return nil, status.Error(codes.InvalidArgument, "`Data` must be provided")
```

lrettig commented on api/grpcserver/smesher_service.go:
```diff
@@ -125,31 +119,112 @@ func (s SmesherService) SetMinGas(ctx context.Context, in *pb.SetMinGasRequest)
+		Throttle:          in.Data.Throttle,
+		ComputeProviderID: int(in.Data.ProviderId),
+	}); err != nil {
+		return nil, err
```
Same comment as above, please wrap the error

lrettig commented on api/grpcserver/smesher_service.go:
```diff
@@ -125,31 +119,112 @@ func (s SmesherService) SetMinGas(ctx context.Context, in *pb.SetMinGasRequest)
-	return nil, status.Errorf(codes.Unimplemented, "this endpoint is not implemented")
+
+	if err := s.post.StopPostDataCreationSession(in.DeleteFiles); err != nil {
+		return nil, err
```
Same comment as above, please wrap the error

lrettig reviewed (dismissed):
Only reviewed the API changes. See inline comments. Also noticed that you added a few TODO items, it would be nice if they were linked to open issues if they remain when the code is merged.

lrettig commented:
Some of this is going to cause conflicts with #2131, which totally removes the old API code. Consider leaving these changes out of this PR, as that one will blow away this code entirely.

moshababo commented:
@lrettig thanks for reviewing. 

---

> Only reviewed the API changes. See inline comments. Also noticed that you added a few TODO items, it would be nice if they were linked to open issues if they remain when the code is merged.

They will be solved prior to merging. 

> Some of this is going to cause conflicts with #2131, which totally removes the old API code. Consider leaving these changes out of this PR, as that one will blow away this code entirely.

I've made changes only to avoid compile-time and test errors. If you expect #2131 to be merged soon, i'll wait and then solve conflicts.

moshababo committed:
Test fixes

moshababo committed:
PR fixes

lrettig committed:
Add link to new issue for reference

lrettig commented:
> If you expect #2131 to be merged soon, i'll wait and then solve conflicts.

I'm still working on it, but I expect it to be ready to be merged by next week.

moshababo committed:
Merge branch 'var_post' into smesher_api (WIP)

moshababo committed:
Resolve conflicts (WIP)

moshababo committed:
WIP

moshababo committed:
Smesher API adjustments (unit tests are broken)

avive commented:
The implementation of the pos progress status stream throws an exception if a pos init process is not currently in progress. The desirable pattern for clients is to subscribe to events on this stream _before_ setting up pos so they can get all errors and status on it while it is ongoing. The stream should be closed by client at any time but never by node (except maybe when running out of resources) . Expecting users to subscribe to stream only _after_ starting a pos init job is not a good pattern because they can miss error events on the stream this way and this is not the normal usage pattern of grpc streaming apis. In addition, I noticed that the stream by the node when a pos setup is complete - this also shouldn't be done and any future pos operation should send data over the stream before the client closed it. I think this is the most straightforward way to implement this kind of functionality.

moshababo committed:
Smesher API adjustments

moshababo commented:
> The implementation of the pos progress status stream throws an exception if a pos init process is not currently in progress. The desirable pattern for clients is to subscribe to events on this stream before setting up pos so they can get all errors and status on it while it is ongoing. 

Fixed. (also see https://github.com/spacemeshos/cli-wallet/pull/23#issuecomment-855256532)

> The stream should be closed by client at any time but never by node (except maybe when running out of resources)

> In addition, I noticed that the stream [IS CLOSED] by the node when a pos setup is complete - this also shouldn't be done and any future pos operation should send data over the stream before the client closed it. I think this is the most straightforward way to implement this kind of functionality.

I’d rather finalize the stream functionality in a follow-up PR, and focus now on the finishing the general API design and implementation.

avive commented:
ok

avive commented:
Building this branch fails on Windows 10 pro with a gcc linker error where building developer branch builds just fine. Likely a GCC issue due to differences between behavior on unix and windows systems. Env: PowerShell / Mingw64.

moshababo committed:
API adjustments, config support

moshababo committed:
Merge branch 'develop' into smesher_api

# Conflicts:
#	.github/workflows/ci.yml
#	DockerFileTests
#	Dockerfile
#	Makefile
#	README.md
#	activation/activation.go
#	activation/activation_test.go
#	activation/activationdb_test.go
#	activation/atxdb.go
#	activation/fetch_mock.go
#	activation/nipst.go
#	activation/nipst_test.go
#	activation/poetdb.go
#	activation/poetlistener_test.go
#	activation/validator.go
#	api/api_test.go
#	api/grpcserver/debug_service.go
#	api/grpcserver/gateway_service.go
#	api/grpcserver/globalstate_service.go
#	api/grpcserver/grpcserver_test.go
#	api/grpcserver/http_server.go
#	api/grpcserver/mesh_service.go
#	api/grpcserver/node_service.go
#	api/grpcserver/smesher_service.go
#	api/node.go
#	blocks/blockeligibilityvalidator.go
#	blocks/blockeligibilityvalidator_test.go
#	blocks/blockoracle.go
#	blocks/blockoracle_test.go
#	blocks/blocks.go
#	blocks/blocks_test.go
#	cmd/hare/oracle.go
#	cmd/node/app_test.go
#	cmd/node/multi_node.go
#	cmd/node/node.go
#	cmd/node/node_test.go
#	cmd/node/oracle.go
#	cmd/p2p/p2p.go
#	cmd/root.go
#	cmd/sync/sync.go
#	cmd/sync/sync_test.go
#	collector/memory_collector.go
#	common/types/activation.go
#	common/types/block.go
#	config/config.go
#	eligibility/fixedoracle.go
#	eligibility/fixedoracle_test.go
#	fetch/fetch.go
#	fetch/fetch_test.go
#	go.mod
#	go.sum
#	hare/algorithm.go
#	hare/algorithm_test.go
#	hare/broker.go
#	hare/eligibility/beacon_test.go
#	hare/eligibility/oracle.go
#	hare/eligibility/oracle_test.go
#	hare/flows_test.go
#	hare/hare.go
#	hare/messagevalidation.go
#	hare/messagevalidation_test.go
#	hare/mock_oracle.go
#	hare/mock_oracle_test.go
#	hare/preroundtracker.go
#	hare/preroundtracker_test.go
#	hare/proposaltracker.go
#	hare/refcounttracker.go
#	layerfetcher/layers.go
#	layerfetcher/layers_test.go
#	log/log.go
#	mesh/mesh.go
#	mesh/mesh_test.go
#	mesh/meshdb.go
#	mesh/meshdb_test.go
#	mesh/reward_test.go
#	miner/block_builder.go
#	miner/builder_test.go
#	p2p/connectionpool/connectionpool_test.go
#	p2p/discovery/getaddress.go
#	p2p/discovery/ping.go
#	p2p/gossip/protocol.go
#	p2p/net/network.go
#	p2p/net/udp.go
#	p2p/server/msgserver.go
#	p2p/server/msgserver_test.go
#	p2p/service/gossip_listener.go
#	p2p/service/gossip_listener_test.go
#	p2p/service/service.go
#	p2p/service/sim.go
#	p2p/switch.go
#	p2p/switch_test.go
#	state/processor.go
#	sync/block_listener.go
#	sync/block_listener_test.go
#	sync/create_baseline_test.go
#	sync/fetch_queue.go
#	sync/fetch_queue_test.go
#	sync/handler.go
#	sync/mocks.go
#	sync/requests.go
#	sync/syncer.go
#	sync/syncer_test.go
#	sync/validation_queue.go
#	sync/worker.go
#	sync/worker_test.go
#	tests/Pipfile
#	tests/README.md
#	tests/block_atx/add_node/config.yaml
#	tests/block_atx/remove_node/config.yaml
#	tests/config.py
#	tests/config.yaml
#	tests/conftest.py
#	tests/deployment.py
#	tests/elk/elastic/elastic_logging.yaml
#	tests/elk/elastic/elastic_stateful.yaml
#	tests/elk/fluent-bit/config.yaml
#	tests/elk/fluent-bit/fluentbit_daemonset.yaml
#	tests/elk/kibana/kibana_logging.yaml
#	tests/elk/logstash/pipeline-config.yml
#	tests/elk/logstash/statefulset.yaml
#	tests/k8s_handler.py
#	tests/late_nodes/delayed_config.yaml
#	tests/node_pool_deployer.py
#	tests/p2p/bootstrap_test/config.yaml
#	tests/p2p/test_p2p.py
#	tests/stress/blocks_stress/config.yaml
#	tests/stress/grpc_stress/config.yaml
#	tests/stress/tx_stress/config.yaml
#	tests/sync/config.yaml
#	tests/sync/test_sync.py
#	tests/tx_generator/config.yaml
#	tests/utils.py
#	timesync/ticker.go
#	tortoise/util.go
#	tortoise/verifying_algorithm.go
#	tortoise/verifying_tortoise.go
#	tortoise/verifying_tortoise_test.go

moshababo committed:
post-merge fixes

sudachen committed:
merged changes

sudachen committed:
updated golang version

sudachen committed:
gofmt

sudachen committed:
updated go.mod

sudachen committed:
updated check-go-version script

sudachen committed:
updated default Makefile target

moshababo committed:
post-merge fixes

moshababo committed:
post-merge fixes

moshababo committed:
Merge branch 'develop' into smesher_api

# Conflicts:
#	activation/activation.go
#	activation/activation_test.go
#	activation/atxdb.go
#	cmd/node/node.go
#	sync/create_baseline_test.go
#	sync/mocks.go

moshababo committed:
post-merge fixes

noamnelke commented on common/types/activation.go:
```diff
@@ -82,13 +82,13 @@ func CalcAtxHash32Presorted(sortedView []ATXID, additionalBytes []byte) Hash32 {
 	id       *ATXID // non-exported cache of the ATXID
-	Space    uint64
 	Coinbase Address
+	NumUnits uint
```
I think it's better to use `uint64` as the header ends up on the wire and this makes the encoding more explicit (this is assuming that `uint`s are encoded as 64-bit unsigned integers anyway and not in an architecture-dependent way, which would be even worse).

moshababo replied:
Yes, already wrote it down for a follow-up PR. Thanks for reminding.

sudachen committed:
added docker_hub option

sudachen committed:
fixed found issues

sudachen committed:
updated Makefile

sudachen committed:
Merge branch 'smesher_api' into playground-gpu-post

sudachen committed:
updated Pipfile.lock file

noamnelke committed:
remove redundant conversion to units

noamnelke committed:
fix overlooked renames

sudachen committed:
uodated system tests configuration

sudachen committed:
fixed panic on PoET get membership

sudachen committed:
removed unsused part from Makefile.Inc

sudachen committed:
Fix panic on PoET get membership

moshababo committed:
Merge branch 'develop' into smesher_api

# Conflicts:
#	activation/activation.go
#	activation/atxdb.go
#	cmd/node/app_test.go
#	cmd/node/multi_node.go
#	cmd/node/node.go
#	config/config.go

moshababo committed:
Merge remote-tracking branch 'origin/smesher_api' into smesher_api

# Conflicts:
#	cmd/node/node.go

moshababo committed:
post-merge fixes

noamnelke committed:
Remove genesis-total-weight (#2518)

Squashed commit of the following:

commit e071e911709885aa1b24fae967413e41f529db59
Merge: cbf06237 45499f4b
Author: Noam Nelke <noam@spacemesh.io>
Date:   Thu Jul 8 21:45:39 2021 +0300

    Merge remote-tracking branch 'origin/develop' into remove-genesis-total-weight

commit cbf06237d9c056d810ad6eb84a3a687fc50e5fcc
Author: Lane Rettig <lanerettig@gmail.com>
Date:   Wed Jul 7 16:27:36 2021 -0400

    Reduce warning to info during genesis

commit a90f75a6de01ba6f01f8d795a4fd48162e8f99c2
Author: Noam Nelke <noam@spacemesh.io>
Date:   Wed Jul 7 15:32:44 2021 +0300

    fix merged system test

commit 733fe08502de71b0fad68e6a9ad6b02cfc64d437
Merge: 7ef18d11 662420dc
Author: Noam Nelke <noam@spacemesh.io>
Date:   Wed Jul 7 13:07:59 2021 +0300

    Merge remote-tracking branch 'origin/develop' into remove-genesis-total-weight

commit 7ef18d11ad7846646c49fb61f4e2789dc710d3d9
Author: Noam Nelke <noam@spacemesh.io>
Date:   Wed Jul 7 12:53:11 2021 +0300

    remove unused total weight calc func

commit 1e26d28d9da1e1121b28af3ac5c07ad7446675eb
Author: Noam Nelke <noam@spacemesh.io>
Date:   Tue Jul 6 21:32:20 2021 +0300

    remove genesis-total-weight

nkryuchkov committed:
Fix tortoise beacon system test in #2518 (#2543)

* Wait after epoch start

# Conflicts:
#	tortoisebeacon/tortoise_beacon.go

* Set epoch weight to 1 if it's 0

* Remove genesis-total-weight flag from tortoise beacon system test

* PR fixes

Co-authored-by: Noam Nelke <noam@spacemesh.io>

sudachen committed:
Merge branch 'smesher_api' into playground-gpu-post

sudachen committed:
updated go.mod

moshababo committed:
Rename: PoST -> Post

moshababo committed:
Merge branch 'develop' into smesher_api

# Conflicts:
#	activation/activation.go
#	activation/activation_test.go
#	activation/atxdb.go
#	activation/nipst.go
#	activation/nipst_test.go
#	activation/poetdb.go
#	cmd/node/multi_node.go
#	cmd/node/node.go
#	cmd/root.go
#	config/config.go
#	go.mod
#	hare/eligibility/oracle.go
#	hare/eligibility/oracle_test.go

moshababo committed:
post-merge fixes

moshababo committed:
activation tests fixes

moshababo committed:
Merge remote-tracking branch 'origin/smesher_api' into smesher_api

moshababo committed:
poet ref fix

moshababo committed:
Update API ref

sudachen committed:
Merge branch 'smesher_api' into playground-gpu-post

sudachen committed:
merged from develop

sudachen committed:
updated DockerfilePrebuiltBinary

moshababo committed:
activation tests fixes

moshababo committed:
Merge branch 'develop' into smesher_api

# Conflicts:
#	activation/activation.go
#	activation/activation_test.go
#	activation/activationdb_test.go
#	activation/atxdb.go
#	api/grpcserver/grpcserver_test.go
#	api/grpcserver/mesh_service.go
#	blocks/blockoracle_test.go
#	cmd/node/app_test.go
#	cmd/node/multi_node.go
#	cmd/node/node.go
#	hare/eligibility/oracle.go
#	hare/eligibility/oracle_test.go
#	mesh/reward_test.go

moshababo committed:
post-merge fixes

moshababo committed:
Merge branch 'smesher_api' into playground-gpu-post

# Conflicts:
#	layerfetcher/layers.go
#	syncer/syncer_test.go

moshababo committed:
Merge pull request #2473 from sudachen/playground-gpu-post

updated Makefile & Dockerfiles

moshababo committed:
Fix makefile OS query

moshababo committed:
allow systemtests run (temp)

moshababo commented:
bors try

sudachen committed:
fix OSX rpath issue

sudachen commented:
bors try

sudachen commented:
bors try

sudachen committed:
disabled unittests in ci.yaml

moshababo committed:
Merge branch 'develop' into smesher_api

# Conflicts:
#	activation/activation.go
#	activation/activation_test.go
#	activation/nipst.go
#	go.mod
#	go.sum

moshababo committed:
post-merge fixes

noamnelke committed:
logs should be dumped when more than one test fails

noamnelke committed:
Merge remote-tracking branch 'origin/develop' into smesher_api

noamnelke committed:
update tortoise beacon config

noamnelke committed:
fix capitaliztion

noamnelke committed:
pass stage1 in ci if UT disabled

noamnelke committed:
for realz

noamnelke commented:
bors try

noamnelke committed:
last time

noamnelke commented:
bors try

sudachen committed:
added test-* target definition for submodules tests (make test-p2p-node for example)

sudachen committed:
removed unused tags

moshababo committed:
fix cmd/sync build

sudachen committed:
Removed test-* targets for floders which don't contain tests

sudachen committed:
uncommented go-sync

sudachen committed:
hide 'for' command in print-test-targets target

sudachen committed:
Added recursive targets test-*-all. Targets test-* is not recursive now and runs tests in the top-level submodule folder. Targets test-*-* are recursive

sudachen committed:
Recursive test-*-all target is defined now for every submodule folder contaning tests

sudachen committed:
Refactored

sudachen committed:
Refactored

sudachen committed:
Fixed print-test-targets, now it uses -all instead [-all] for folders which have no own tests

sudachen committed:
ulimit is depended on host system now

sudachen committed:
Refactored

noamnelke committed:
update p2p test clientv2 image

noamnelke committed:
ci tweak

noamnelke commented:
bors try

noamnelke committed:
fmt

noamnelke committed:
some easy unit test fixes

moshababo committed:
Merge branch 'develop' into smesher_api

noamnelke committed:
temporarily remove unit tests from ci

noamnelke commented:
bors try

moshababo committed:
test fixes

moshababo committed:
Merge remote-tracking branch 'origin/smesher_api' into smesher_api

moshababo committed:
test fixes

noamnelke committed:
restore ci.yml

moshababo committed:
lint fix

moshababo committed:
Merge remote-tracking branch 'origin/smesher_api' into smesher_api

moshababo committed:
Update p2p/switch.go

moshababo committed:
go mod tidy

moshababo committed:
go mod tidy

moshababo committed:
Makefile: fix exclude_app_test flag

moshababo committed:
test fixes

moshababo committed:
Update API release

moshababo committed:
debug: reduce Post params difficulty

moshababo committed:
temporarily disable `TestMultiNode`

moshababo commented:
bors try

moshababo committed:
Code cleanup

moshababo committed:
Code cleanup

moshababo committed:
Update `spacemeshos/post` dependency target commit

moshababo committed:
Update gitignore

moshababo committed:
go mod tidy

moshababo commented:
bors merge

moshababo commented:
bors merge

moshababo commented:
bors merge
